### PR TITLE
Feature/ball placement us play

### LINKED ISSUE #1009: Possibility to avoid ball

### PR DESCRIPTION
**Pull request by RobotJente**
Made ball placement play.

Does not currently optimally interact with the referee commands, as the ball-placement can be followed with 2 different commands: [ssl-ball-placement-rules](https://robocup-ssl.github.io/ssl-rules/sslrules.html#_ball_placement)
However the current play does follow the rules

### CODE REVIEW

File: include/roboteam_ai/stp/StpInfo.h [hunk 1]
```diff
@@ -175,6 +179,13 @@ struct StpInfo {
      * Used to decide how close this robot should be to enemy robot
      */
     BlockDistance blockDistance;
+
+    /**
+     * When avoiding the ball, the robot will try to keep this distance (in meters) between the ball and the robot)
+     */
+    std::optional<double> avoidBallDistance;
+
+
```
Comment (RobotJente) [inline include/roboteam_ai/stp/StpInfo.h]:
```suggestion
```

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 2]
```diff
@@ -0,0 +1,61 @@
...
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_4")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_5")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_6")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_7")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_8")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_9")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_10"))};
+    }
+
+    uint8_t BallPlacementThem::score(world_new::World* world) noexcept { return 0; }
+
+    void BallPlacementThem::calculateInfoForRoles() noexcept {
+
```
Comment (RobotJente) [inline src/stp/new_plays/BallPlacementThem.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/StpInfo.h [hunk 3]
```diff
@@ -175,6 +179,12 @@ struct StpInfo {
      * Used to decide how close this robot should be to enemy robot
      */
     BlockDistance blockDistance;
+
+    /**
+     * When avoiding the ball, the robot will try to keep this distance (in meters) between the ball and the robot)
+     */
+    std::optional<double> avoidBallDistance;
+
```
Comment (RobotJente) [inline include/roboteam_ai/stp/StpInfo.h]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays/BallPlacementUs.h [hunk 4]
```diff
@@ -1,11 +1,15 @@
+#ifndef RTT_BALLPLACEMENTUS_H
+#define RTT_BALLPLACEMENTUS_H
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/BallPlacementUs.h:2]: How does this .h  file work without including Play.h?

File: include/roboteam_ai/stp/new_plays/BallPlacementThem.h [hunk 5]
```diff
@@ -0,0 +1,44 @@
+#include "stp/Play.hpp"
+#ifndef RTT_BALLPLACEMENTTHEM_H
+#define RTT_BALLPLACEMENTTHEM_H
+namespace rtt::ai::stp::play {
+
+    class BallPlacementThem : public Play {
```
Comment (timovdk) [inline include/roboteam_ai/stp/new_plays/BallPlacementThem.h]:
```suggestion
#ifndef RTT_BALLPLACEMENTTHEM_H
#define RTT_BALLPLACEMENTTHEM_H

#include "stp/Play.hpp"

namespace rtt::ai::stp::play {

    class BallPlacementThem : public Play {
```

File: src/ApplicationManager.cpp [hunk 6]
```diff
@@ -13,7 +13,7 @@
 #include "stp/new_plays/Defend.h"
 #include "stp/new_plays/Attack.h"
 #include "stp/new_plays/Halt.h"
-#include "stp/new_plays/BallPlacement.h"
+#include "stp/new_plays/BallPlacementUs.h"
```
Comment (timovdk) [inline src/ApplicationManager.cpp:16]: You did not include BallPlacementThem yet

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 7]
```diff
@@ -0,0 +1,60 @@
+//
+// Created by jessevw on 24.03.20.
+//
+
+#include "stp/new_plays/Halt.h"
```
Comment (timovdk) [inline src/stp/new_plays/BallPlacementThem.cpp]: You don't need to include the halt play

File: include/roboteam_ai/stp/StpInfo.h [hunk 8]
```diff
@@ -114,6 +114,10 @@ struct StpInfo {
     KickChipType getKickChipType() const { return kickChipType; }
     void setKickChipType(KickChipType kickChipType) { this->kickChipType = kickChipType; }
 
+    const std::optional<double> &getAvoidBallDistance() const { return avoidBallDistance; }
+    void setAvoidBallDistance(const std::optional<double> &avoidBallDistance) { StpInfo::avoidBallDistance = avoidBallDistance; }
```
Comment (jordiluong) [inline include/roboteam_ai/stp/StpInfo.h]:
```suggestion
    void setAvoidBallDistance(const std::optional<double> &avoidBallDistance) { this->avoidBallDistance = avoidBallDistance; }
```

File: src/stp/new_tactics/AvoidBall.cpp [hunk 9]
```diff
@@ -0,0 +1,66 @@
+//
+// Created by jesse on 30-04-20.
+//
+
+#include <include/roboteam_ai/utilities/GameStateManager.hpp>
```
Comment (timovdk) [inline src/stp/new_tactics/AvoidBall.cpp]:
```suggestion
#include <utilities/GameStateManager.hpp>
```

File: src/stp/new_tactics/AvoidBall.cpp [hunk 10]
```diff
@@ -0,0 +1,66 @@
...
+        Vector2 stretchedBallToRobot;
+        if (info.getAvoidBallDistance()) {
+            stretchedBallToRobot = ballToRobot.stretchToLength(info.getAvoidBallDistance().value());
+        }
+        else {
+            stretchedBallToRobot = ballToRobot.stretchToLength(0.5);
+        }
+
+        auto positionFromBall = info.getRobot()->get()->getPos() + stretchedBallToRobot;
+        //skillStpInfo.setPositionToMoveTo(positionFromBall);
```
Comment (timovdk) [inline src/stp/new_tactics/AvoidBall.cpp]: Commented out code

File: src/ApplicationManager.cpp (hunk 6 above)
Comment (jordiluong) [inline src/ApplicationManager.cpp:16]:
```suggestion
#include "stp/new_plays/BallPlacementUs.h"
#include "stp/new_plays/BallPlacementThem.h"
```

File: src/ApplicationManager.cpp [hunk 11]
```diff
@@ -45,7 +45,7 @@ void ApplicationManager::start() {
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::Defend>());
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::DefensiveFormation>());
     plays.emplace_back(std::make_unique<rtt::ai::stp::play::AggressiveFormation>());
-    plays.emplace_back(std::make_unique<rtt::ai::stp::play::BallPlacement>());
+    plays.emplace_back(std::make_unique<rtt::ai::stp::play::BallPlacementUs>());
```
Comment (jordiluong) [inline src/ApplicationManager.cpp:49]:
```suggestion
    plays.emplace_back(std::make_unique<rtt::ai::stp::play::BallPlacementUs>());
    plays.emplace_back(std::make_unique<rtt::ai::stp::play::BallPlacementThem>());
```

File: src/roboteam_ai.cpp [hunk 12]
```diff
@@ -75,7 +75,7 @@ int main(int argc, char* argv[]) {
     rtt::SETTINGS.setVisionIp("127.0.0.1");
     rtt::SETTINGS.setVisionPort(10006);
     rtt::SETTINGS.setRefereeIp("224.5.23.1");
-    rtt::SETTINGS.setRefereePort(10007);
+    rtt::SETTINGS.setRefereePort(10003);
```
Comment (jordiluong) [inline src/roboteam_ai.cpp:78]: Don't know if you should change this or not. Mine works at 10007?

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 13]
```diff
@@ -0,0 +1,60 @@
...
+#include "stp/invariants/game_states/BallPlacementThemGameStateInvariant.h"
+#include "stp/new_roles/BallPlacer.h"
+#include "stp/new_roles/BallAvoider.h"
+#include "stp/new_plays/BallPlacementThem.h"
+
+namespace rtt::ai::stp::play {
+
+    BallPlacementThem::BallPlacementThem() : Play() {
+        // TODO: decide start invariants
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementThem.cpp]:
```suggestion
```

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 14]
```diff
@@ -0,0 +1,60 @@
...
+        startPlayInvariants.clear();
+        startPlayInvariants.emplace_back(std::make_unique<invariant::BallPlacementThemGameStateInvariant>());
+
+        // TODO: decide keep invariants
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementThem.cpp]:
```suggestion
```

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 15]
```diff
@@ -0,0 +1,60 @@
...
+
+    void BallPlacementThem::calculateInfoForRoles() noexcept {
+    }
+
+    bool BallPlacementThem::shouldRoleSkipEndTactic() { return false; }
+
+    Dealer::FlagMap BallPlacementThem::decideRoleFlags() const noexcept {
+        Dealer::FlagMap flagMap;
+        Dealer::DealerFlag closeToBallFlag(DealerFlagTitle::CLOSE_TO_BALL, DealerFlagPriority::HIGH_PRIORITY);
+        Dealer::DealerFlag notImportant(DealerFlagTitle::CLOSE_TO_OUR_GOAL, DealerFlagPriority::LOW_PRIORITY);
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementThem.cpp]:
```suggestion
```

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 16]
```diff
@@ -0,0 +1,60 @@
...
+        flagMap.insert({"ball_avoider_1", {notImportant}});
+        flagMap.insert({"ball_avoider_2", {notImportant}});
+        flagMap.insert({"ball_avoider_3", {notImportant}});
+        flagMap.insert({"ball_avoider_4", {notImportant}});
+        flagMap.insert({"ball_avoider_5", {notImportant}});
+        flagMap.insert({"ball_avoider_6", {notImportant}});
+        flagMap.insert({"ball_avoider_7", {notImportant}});
+        flagMap.insert({"ball_avoider_8", {notImportant}});
+        flagMap.insert({"ball_avoider_9", {notImportant}});
+        flagMap.insert({"ball_avoider_10", {notImportant}});
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementThem.cpp]:
```suggestion
        flagMap.insert({"ball_avoider_0", {}});
        flagMap.insert({"ball_avoider_1", {}});
        flagMap.insert({"ball_avoider_2", {}});
        flagMap.insert({"ball_avoider_3", {}});
        flagMap.insert({"ball_avoider_4", {}});
        flagMap.insert({"ball_avoider_5", {}});
        flagMap.insert({"ball_avoider_6", {}});
        flagMap.insert({"ball_avoider_7", {}});
        flagMap.insert({"ball_avoider_8", {}});
        flagMap.insert({"ball_avoider_9", {}});
        flagMap.insert({"ball_avoider_10", {}});
```

File: src/stp/new_plays/BallPlacementUs.cpp [hunk 17]
```diff
@@ -0,0 +1,67 @@
...
+#include "stp/invariants/game_states/BallPlacementUsGameStateInvariant.h"
+#include "stp/new_roles/BallPlacer.h"
+#include "stp/new_roles/BallAvoider.h"
+#include "stp/new_plays/BallPlacementUs.h"
+#include "utilities/GameStateManager.hpp"
+
+namespace rtt::ai::stp::play {
+
+    BallPlacementUs::BallPlacementUs() : Play() {
+        // TODO: decide start invariants
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementUs.cpp]:
```suggestion
```

File: src/stp/new_plays/BallPlacementUs.cpp [hunk 18]
```diff
@@ -0,0 +1,67 @@
...
+        startPlayInvariants.clear();
+        startPlayInvariants.emplace_back(std::make_unique<invariant::BallPlacementUsGameStateInvariant>());
+
+        // TODO: decide keep invariants
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementUs.cpp]:
```suggestion
```

File: include/roboteam_ai/stp/new_plays/BallPlacementUs.h (hunk 4 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/new_plays/BallPlacementUs.h:2]: Its included in the cpp

File: src/stp/new_plays/BallPlacementUs.cpp [hunk 19]
```diff
@@ -0,0 +1,67 @@
...
+            RTT_ERROR("No ball placement robot assigned!")
+        }
+    }
+
+    bool BallPlacementUs::shouldRoleSkipEndTactic() { return false; }
+
+    Dealer::FlagMap BallPlacementUs::decideRoleFlags() const noexcept {
+        Dealer::FlagMap flagMap;
+        Dealer::DealerFlag closeToBallFlag(DealerFlagTitle::CLOSE_TO_BALL, DealerFlagPriority::HIGH_PRIORITY);
+        Dealer::DealerFlag notImportant(DealerFlagTitle::CLOSE_TO_OUR_GOAL, DealerFlagPriority::LOW_PRIORITY);
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementUs.cpp]:
```suggestion
```

File: src/stp/new_plays/BallPlacementUs.cpp [hunk 20]
```diff
@@ -0,0 +1,67 @@
...
+        flagMap.insert({"ball_avoider_1", {notImportant}});
+        flagMap.insert({"ball_avoider_2", {notImportant}});
+        flagMap.insert({"ball_avoider_3", {notImportant}});
+        flagMap.insert({"ball_avoider_4", {notImportant}});
+        flagMap.insert({"ball_avoider_5", {notImportant}});
+        flagMap.insert({"ball_avoider_6", {notImportant}});
+        flagMap.insert({"ball_avoider_7", {notImportant}});
+        flagMap.insert({"ball_avoider_8", {notImportant}});
+        flagMap.insert({"ball_avoider_9", {notImportant}});
+        flagMap.insert({"ball_avoider_10", {notImportant}});
```
Comment (jordiluong) [inline src/stp/new_plays/BallPlacementUs.cpp]:
```suggestion
        flagMap.insert({"ball_avoider_1", {}});
        flagMap.insert({"ball_avoider_2", {}});
        flagMap.insert({"ball_avoider_3", {}});
        flagMap.insert({"ball_avoider_4", {}});
        flagMap.insert({"ball_avoider_5", {}});
        flagMap.insert({"ball_avoider_6", {}});
        flagMap.insert({"ball_avoider_7", {}});
        flagMap.insert({"ball_avoider_8", {}});
        flagMap.insert({"ball_avoider_9", {}});
        flagMap.insert({"ball_avoider_10", {}});
```

**Review by jordiluong — CHANGES_REQUESTED** [review]

File: include/roboteam_ai/stp/new_plays/BallPlacementUs.h (hunk 4 above)
Comment (timovdk) [reply include/roboteam_ai/stp/new_plays/BallPlacementUs.h:2]: But everywhere else we include the parent class in the .h file

File: src/roboteam_ai.cpp (hunk 12 above)
Comment (timovdk) [reply src/roboteam_ai.cpp:78]: mine at 10003 as well

File: include/roboteam_ai/stp/new_plays/BallPlacementUs.h (hunk 4 above)
Comment (RobotJente) [reply include/roboteam_ai/stp/new_plays/BallPlacementUs.h:2]: Fixed it :)

File: include/roboteam_ai/stp/new_tactics/AvoidBall.h [hunk 21]
```diff
@@ -0,0 +1,63 @@
+//
+// Created by jordi on 08-04-20.
```
Comment (RobotJente) [inline include/roboteam_ai/stp/new_tactics/AvoidBall.h]:
```suggestion
// Created by RobotJesse on 08-04-20.
```

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 22]
```diff
@@ -0,0 +1,55 @@
...
+        keepPlayInvariants.emplace_back(std::make_unique<invariant::BallPlacementThemGameStateInvariant>());
+
+        roles = std::array<std::unique_ptr<Role>, rtt::ai::Constants::ROBOT_COUNT()>{
+                std::make_unique<role::BallPlacer>(role::BallPlacer("ball_avoider_0")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_1")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_2")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_3")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_4")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_5")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_6")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_7")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_8")),    std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_9")),
+                std::make_unique<role::BallAvoider>(role::BallAvoider("ball_avoider_10"))};
+    }
```
Comment (timovdk) [inline src/stp/new_plays/BallPlacementThem.cpp:26]: You probably also want to assign a keeper in this play

File: src/stp/new_plays/BallPlacementThem.cpp [hunk 23]
```diff
@@ -0,0 +1,55 @@
...
+        flagMap.insert({"ball_avoider_1", {}});
+        flagMap.insert({"ball_avoider_2", {}});
+        flagMap.insert({"ball_avoider_3", {}});
+        flagMap.insert({"ball_avoider_4", {}});
+        flagMap.insert({"ball_avoider_5", {}});
+        flagMap.insert({"ball_avoider_6", {}});
+        flagMap.insert({"ball_avoider_7", {}});
+        flagMap.insert({"ball_avoider_8", {}});
+        flagMap.insert({"ball_avoider_9", {}});
+        flagMap.insert({"ball_avoider_10", {}});
```
Comment (timovdk) [inline src/stp/new_plays/BallPlacementThem.cpp:48]: And add the not_important flags

File: src/stp/new_plays/BallPlacementUs.cpp [hunk 24]
```diff
@@ -0,0 +1,69 @@
...
+
+    uint8_t BallPlacementUs::score(world_new::World* world) noexcept { return 100; }
+
+    void BallPlacementUs::calculateInfoForRoles() noexcept {
+        if (stpInfos.find("ball_placer") != stpInfos.end()) {
+            auto ballTarget = rtt::ai::GameStateManager::getRefereeDesignatedPosition();
+            stpInfos["ball_placer"].setPositionToMoveTo(ballTarget);
+        }
+        if (stpInfos.find("keeper") != stpInfos.end()) {
+            stpInfos["keeper"].setEnemyRobot(world->getWorld()->getRobotClosestToBall(world_new::Team::them));
```
Comment (timovdk) [inline src/stp/new_plays/BallPlacementUs.cpp:37]: I think the keeper also needs a positiontoshootat for it to not segfault

File: src/stp/new_plays/BallPlacementThem.cpp (hunk 22 above)
Comment (RobotJente) [reply src/stp/new_plays/BallPlacementThem.cpp:26]: yup made an issue for it

File: src/stp/new_plays/BallPlacementThem.cpp (hunk 23 above)
Comment (RobotJente) [reply src/stp/new_plays/BallPlacementThem.cpp:48]: Made an issue for this as well

File: src/stp/new_plays/BallPlacementUs.cpp (hunk 24 above)
Comment (RobotJente) [reply src/stp/new_plays/BallPlacementUs.cpp:37]: Oh okay I will update the issue for this as well

**Review by timovdk — APPROVED** [review]

**Pull request merged**